core/src/main/java/com/exadel/kaliada/core/services/impl/HarvardNewsSearchServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+package com.exadel.kaliada.core.services.impl;
+
+import com.day.cq.tagging.Tag;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.PageManager;
+import com.exadel.kaliada.core.models.NewsModel;
+import com.exadel.kaliada.core.services.HarvardNewsSearchService;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.ResourceResolver;
+import org.jsoup.Jsoup;
+import org.osgi.service.component.annotations.Component;
+
+import javax.jcr.Node;
+import javax.jcr.NodeIterator;
+import javax.jcr.RepositoryException;
+import javax.jcr.Session;
+import javax.jcr.query.Query;
+import javax.jcr.query.QueryManager;
+import javax.jcr.query.QueryResult;
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+@Component(service = HarvardNewsSearchService.class, immediate = true)
+public class HarvardNewsSearchServiceImpl implements HarvardNewsSearchService {
+    private static final String BASE_QUERY = "SELECT * FROM [cq:PageContent] AS s WHERE ISDESCENDANTNODE([/content/aemdemo/locale]) AND s.[cq:template]='/conf/aemdemo/settings/wcm/templates/harvard-single-news-page' AND s.[cq:tags] LIKE 'harvard-news:%' order by s.[jcr:created] desc";
+    private static final int COUNT_WORDS_SUMMARY_ARTICLE = 200;
+    private static final String LOCALE_NAME_IN_BASE_QUERY = "locale";
+    private static final String TITLE_NODE = "root/container/title";
+    private static final String TITLE_PROPERTY = "jcr:title";
+    private static final String TAG_PROPERTY = "cq:tags";
+    private static final String TAG_NAME_SPACE = "harvard-news:";
+    private static final String IMAGE_NODE = "root/container/image";
+    private static final String IMAGE_PROPERTY = "fileReference";
+    private static final String TEXT_NODE = "root/container/text";
+    private static final String TEXT_PROPERTY = "text";
+    private static final String LIKE_NODE = "root/container/like";
+    private static final String LIKE_PROPERTY = "likeCounter";
+    private static final String DISLIKE_PROPERTY = "dislikeCounter";
+    private static final String ROOT_PAGE = "/content/aemdemo/en";
+    private static final String EXTENSION = ".html";
+    private static final String TAG_URL_PARAMETER = "?tag=";
+    private static final String URI_HARVARD_NEWS = "/content/aemdemo/en/harvard-news.html";
+
+    @Override
+    public List<NewsModel> getAllNews(SlingHttpServletRequest request, int offset, int limit, String tagName, String locale) {
+        try {
+            List<NewsModel> news = new ArrayList<>();
+            Query query = getQuery(request, tagName, locale, offset, limit);
+            QueryResult queryResult = query.execute();
+            NodeIterator nodeIterator = queryResult.getNodes();
+            while (nodeIterator.hasNext()) {
+                Node node = nodeIterator.nextNode();
```

Comment:
I suggest using `modelFactory.getModelFromWrappedRequest()` for such cases. So that models are still injected, not created like a simple POJO.
And please, always prefer operating with Resource tier, not Node tier. resource.getValueMap().get("foo", 0) ... .get("bar", String.class) is a friend of yours)

Reply:
fixed, thanks